build: Fix static Darwin builds

hebasto opened the pull request:
From `##bitcoin-core-gui` IRC channel:

> \<fanquake> hebasto: I think I might have solved the darkmode issue, or at least some of the macOS qt problems. Will PR shortly
> \<hebasto> was I on the right way?
> \<fanquake> Yes I think it's related to how the SDK is being setup
> \<hebasto> fanquake: expecting of adding `-Wl,-sdk_version` to linker flags :)
> \<fanquake> hebasto: heh. That may be added at some point, however I think the solution here will be even simpler

Assuming, that **fanquake**'s solution does not involve linker flags, I'm suggesting an alternative one that does use them.

When doing static builds, for some reasons, SDK version is not handled correctly when Darwin Driver [passes](https://github.com/llvm/llvm-project/commit/25ce33a6e4f3b13732c0f851e68390dc2acb9123) it to a linker via the
`-platform_version` option.

With this PR we pass the `-platform_version` option to a linker explicitly.

Fix #21771.

This change causes harmless [warnings](https://github.com/tpoechtrager/cctools-port/blob/6c438753d2252274678d3e0839270045698c159b/cctools/ld64/src/ld/Options.cpp#L3213-L3214):
```
ld: warning: passed two min versions (10.14.0, 10.14) for platform macOS. Using 10.14.
```

---

Building with depends on macOS Big Sur 11.2.3 (20D91):
```
% ld -v
@(#)PROGRAM:ld  PROJECT:ld64-609.8
BUILD 15:07:46 Dec 18 2020
configured to support archs: armv6 armv7 armv7s arm64 arm64e arm64_32 i386 x86_64 x86_64h armv6m armv7k armv7m armv7em
LTO support using: LLVM version 12.0.0, (clang-1200.0.32.29) (static support for 27, runtime is 27)
TAPI support using: Apple TAPI version 12.0.0 (tapi-1200.0.23.5)
```
- on master (683dda2a70e7a210996fa34be23bd0c563971ba9)
```
% otool -l src/qt/bitcoin-qt | grep -A 7 -B 1 BUILD_VERSION
Load command 9
      cmd LC_BUILD_VERSION
  cmdsize 32
 platform 1
    minos 10.14
      sdk n/a
   ntools 1
     tool 3
  version 609.8
```
- with this PR
```
% otool -l src/qt/bitcoin-qt | grep -A 7 -B 1 BUILD_VERSION                       
Load command 9
      cmd LC_BUILD_VERSION
  cmdsize 32
 platform 1
    minos 10.14
      sdk 11.1
   ntools 1
     tool 3
  version 609.8
```

hebasto committed:
build: Fix static Darwin builds

When doing static builds, for some reasons, SDK version is not handled
correctly when Darwin Driver passes it to a linker via the
-platform_version option.

This change makes passing the -platform_version option to a linker
explicit.

RandyMcMillan commented:
Tested on macOS 10.14.6 (18G103)

jarolrod reviewed (commented):
macOS 10.14 SDK doesn't seem to have `-platform_version` available. When trying to perform a native depends build on macOS 10.14 I get the following error when building `libevent`:

```
ld: unknown option: -platform_version
clang: error: linker command failed with exit code 1 (use -v to see invocation)
configure:4267: $? = 1
configure:4307: result: no
```

The above error would mean this PR wouldn't work in its current form.

Besides broken compatibility with 10.14, this PR seems to be a step in the right direction. The following will be based on my testing of the state of dark mode: https://github.com/bitcoin-core/gui/pull/275#pullrequestreview-639461197

#### macOS 10.15

| Master        | PR               |
| ----------- | ----------- |
| ![115327729-6e549700-a15d-11eb-8c46-7c830007d71e](https://user-images.githubusercontent.com/23396902/116174288-1fb67800-a6dc-11eb-8673-2e2a96ee4de1.png) | ![10 15](https://user-images.githubusercontent.com/23396902/116174156-dfef9080-a6db-11eb-805f-9eb8b2f61d23.png) | 

```
otool -l src/qt/bitcoin-qt | grep -A 7 -B 1 BUILD_VERSION  
       
Load command 9
      cmd LC_BUILD_VERSION
  cmdsize 32
 platform 1
    minos 10.14
      sdk 10.15.6
   ntools 1
     tool 3
  version 609.8
```

#### macOS 11.2.3

| Master        | PR               |
| ----------- | ----------- |
| ![115330418-fccb1780-a161-11eb-84ca-1c87a6b5f448](https://user-images.githubusercontent.com/23396902/116174380-4c6a8f80-a6dc-11eb-9215-6afa197ac9cc.png) | ![11 2](https://user-images.githubusercontent.com/23396902/116174393-52607080-a6dc-11eb-91ab-1b5aeffef995.png) |

```
Load command 9
      cmd LC_BUILD_VERSION
  cmdsize 32
 platform 1
    minos 10.14
      sdk 11.1
   ntools 1
     tool 3
  version 609.8
```

hebasto commented:
@RandyMcMillan 
> Tested on macOS 10.14.6 (18G103)

Did you test a static build?

hebasto commented:
Updated 15d41b3b0120d9093036dc3d1740df228be25cc7 -> d7c83cb067992f38d0b6f6d3a7896f971718e24b ([pr21782.01](https://github.com/hebasto/bitcoin/commits/pr21782.01) -> [pr21782.02](https://github.com/hebasto/bitcoin/commits/pr21782.02), [diff](https://github.com/hebasto/bitcoin/compare/pr21782.01..pr21782.02)):

- fixed compatibility with macOS Mojave

RandyMcMillan commented:
For clarity:

`
brew install mingw-w64
`
was required to build depends on Mojave

![Screen Shot 2021-04-27 at 8 01 27 AM](https://user-images.githubusercontent.com/152159/116237938-ccb9e080-a72e-11eb-8de2-ad117ee5bd87.png)
![Screen Shot 2021-04-27 at 8 02 08 AM](https://user-images.githubusercontent.com/152159/116238041-e65b2800-a72e-11eb-8818-e929ac7fc5bd.png)

hebasto commented:
@RandyMcMillan 
> For clarity:
> 
> `brew install mingw-w64`
> was required to build depends on Mojave

Sorry, I don't understand this.

`mingw-w64` is required only when cross-compiling for Windows target.

This pr deals with macOS target only.

RandyMcMillan commented:
Almost got thru it.

```
Darwin ₿ 18.7.0 Darwin Kernel Version 18.7.0: Tue Aug 20 16:57:14 PDT 2019; root:xnu-4903.271.2~2/RELEASE_X86_64 x86_64
```

commit 
d7c83cb067992f38d0b6f6d3a7896f971718e24b

Used commands:
```
$ make -C depends
$ ./autogen.sh
$ CONFIG_SITE=$PWD/depends/x86_64-apple-darwin18.7.0/share/config.site ./configure
$ make
```

![Screen Shot 2021-04-27 at 11 32 12 AM](https://user-images.githubusercontent.com/152159/116270323-fb46b400-a74c-11eb-851b-a4fd44945754.png)

hebasto commented:
@RandyMcMillan 
Could `make clean && make` help?

RandyMcMillan commented:
Error:

![Screen Shot 2021-04-27 at 4 38 09 PM](https://user-images.githubusercontent.com/152159/116311397-35c64600-a779-11eb-864d-b30fc173dff0.png)

Disabled Tests:

```
CONFIG_SITE=$PWD/depends/x86_64-apple-darwin18.7.0/share/config.site ./configure --disable-tests
make
```
![Screen Shot 2021-04-27 at 4 55 15 PM](https://user-images.githubusercontent.com/152159/116311594-69a16b80-a779-11eb-9a98-7f556bf0a298.png)


#### Besides the test error d7c83cb06 seems good - This may have fixed a dbus issue I was seeing as well.

jarolrod commented:
@RandyMcMillan there is most likely something wrong with your Qt installation, d7c83cb builds just fine for me on macOS 10.14

hebasto commented:
@RandyMcMillan 

To figure out problems with this PR, could you confirm a successful build with depends on macOS 10.14?

fanquake commented:
How is [this comment](https://github.com/bitcoin/bitcoin/pull/21782#pullrequestreview-645282708):
> macOS 10.14 SDK doesn't seem to have -platform_version available. When trying to perform a native depends build on macOS 10.14 I get the following error

solved with the current change? If the linker doesn't support `-platform_version`, then building with a depends prefix would still be broken, because `-platform_version` still gets added to `LDFLAGS` when configuring.

@jarolrod It's not the 10.14 SDK that doesn't support `-platform_version`, it's the linker. What version of `ld` are you using where `-platform_version` isn't supported?

jarolrod commented:
@fanquake on my macOS 10.14.6 partition:

```
ld -v

@(#)PROGRAM:ld  PROJECT:ld64-450.3
BUILD 18:16:53 Apr  5 2019
configured to support archs: armv6 armv7 armv7s arm64 arm64e arm64_32 i386 x86_64 x86_64h armv6m armv7k armv7m armv7em
LTO support using: LLVM version 10.0.1, (clang-1001.0.46.4) (static support for 22, runtime is 22)
TAPI support using: Apple TAPI version 10.0.1 (tapi-1001.0.4.1)
```

I failed to build natively on macOS with depends using 15d41b3, but it builds using d7c83cb. Also, I don't see `-platform_version` under the `LDFLAGS` section of my configure output on 10.14.6. I do see it on 10.15 and 11.

fanquake commented:
Right, because `ld` actually throws an error, `AX_CHECK_LINK_FLAG` wont append the flags. I had been thinking it was just warning, and because the call wasn't using `$LDFLAG_WERROR` they would end up appended and then we'd fail to link.

Regardless, as-is, this change is a bit of an undocumented hack. It need explanation either in depends, or in configure.

Just to clarify, as I understand it, static builds aren't "broken", it's just that our current macOS toolchain, whether cross-compiling via depends, or building on macOS, doesn't fill out the `sdk` field in the `LC_BUILD_VERSION` command. This is an issue because Qt is now detecting at runtime, via the version in that field, whether or not to enable some features, such as "Dark mode"?

jarolrod commented:
@fanquake hopefully this can clarify a bit, going by my "state of dark mode" testing https://github.com/bitcoin-core/gui/pull/275#pullrequestreview-639461197

From the linked review, we can see that this `sdk field` issue, seemingly, manifests itself only on native depends builds and with the bumped toolchain cross-compile (did not test bumped toolchain native depends build). 

**Linux Cross Compile**
The sdk field ***is*** filled in when cross-compiling for macOS. Albeit, with a (maybe) strange value for sdk. Anyways, `Dark Mode` is enabled. Below is the output of `otool` on a `bitcoin-qt` binary cross-compiled from linux on master:

```
otool -l src/qt/bitcoin-qt | grep -A 7 -B 1 BUILD_VERSION

Load command 9
      cmd LC_BUILD_VERSION
  cmdsize 32
 platform 1
    minos 10.14
      sdk 505.0
   ntools 1
     tool 3
  version 530.0
```

**Native macOS Depends Build**
On a native macOS depends build, the `SDK` value is not filled in. it registers as `N/A` using `otool`. `Dark Mode` is disabled. The only exception, before this PR, is macOS 10.14 where there is a value for `sdk`.

**Bumped Toolchain Cross Compile**
Building a macOS cross-compile using the bumped toolchain (https://github.com/bitcoin/bitcoin/pull/19817). The `sdk` field is not filled in and `Dark Mode` is disabled. Here is the output of `otool` on a `bitcoin-qt` binary cross-compiled from Linux on the bumped toolchain branch: 

```
bitcoin % otool -l src/qt/bitcoin-qt | grep -A 7 -B 1 BUILD_VERSION

Load command 9
      cmd LC_BUILD_VERSION
  cmdsize 32
 platform 1
    minos 10.14
      sdk n/a
   ntools 1
     tool 3
  version 609.0
```

**Where this PR fits in**
From my testing, this PR fills in the `sdk` field of the affected builds (native depends, and bumped toolchain cross-compile)

Native depends build have the `sdk` field filled in and `Dark Mode` works ✅

I combined the bumped toolchain and this PR on this branch: https://github.com/jarolrod/bitcoin/tree/ctool-and-static-darwin. Bumped Toolchain Cross Compile otool output with this pr:
```
otool -l src/qt/bitcoin-qt | grep -A 7 -B 1 BUILD_VERSION

Load command 9
      cmd LC_BUILD_VERSION
  cmdsize 32
 platform 1
    minos 10.14
      sdk 10.15.6
   ntools 1
     tool 3
  version 609.0
```

Another (welcome) side-effect of this PR is that the `sdk` field is properly filled in on a cross-compile from the current master. `otool` reports back the expected value of `10.15.1` instead of `505.0` that we currently get without this PR. 
```
otool -l src/qt/bitcoin-qt | grep -A 7 -B 1 BUILD_VERSION

Load command 9
      cmd LC_BUILD_VERSION
  cmdsize 32
 platform 1
    minos 10.14
      sdk 10.15.1
   ntools 1
     tool 3
  version 530.0

```

fanquake commented:
> From the linked review, we can see that this sdk field issue, seemingly, manifests itself only on native depends builds and with the bumped toolchain cross-compile (did not test bumped toolchain native depends build).

The SDK field has also been present in the `LC_VERSION_MIN_MACOSX` command; `LC_BUILD_VERSION` is actually fairly new. I'm not convinced this approach is the way to solve our issue. I think the more correct fix is #21793.

hebasto commented:
Closing in favor of #21793 which is more correct solution for #21771.

hebasto commented:
@fanquake 
> The SDK field has also been present in the `LC_VERSION_MIN_MACOSX` command; `LC_BUILD_VERSION` is actually fairly new.

I cannot see `LC_VERSION_MIN_MACOSX` in the `otool` output on macOS.

fanquake commented:
> I cannot see LC_VERSION_MIN_MACOSX in the otool output on macOS.

It's going to depend on which toolchain you've used. The most recently released binaries have `LC_VERSION_MIN_MACOSX` not `LC_BUILD_VERSION`.
